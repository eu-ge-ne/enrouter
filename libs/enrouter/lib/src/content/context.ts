import { createContext, useContext } from "react";

import type { Content } from "./mod.js";

const ContentContext = createContext<Content>({
  match: {
    route: {
      path: "",
      test: {
        keys: [],
        pattern: new RegExp(""),
      },
      modules: [],
      loaded: false,
      elements: {},
    },
    location: "",
    isFull: false,
    params: {},
  },
});

export const ContentProvider = ContentContext.Provider;

export function useContent(): Content {
  return useContext(ContentContext);
}
