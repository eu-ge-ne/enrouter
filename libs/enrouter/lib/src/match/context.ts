import { createContext, useContext } from "react";

import type { Match } from "./mod.js";

const MatchContext = createContext<Match>({
  route: {
    path: "",
    test: { keys: [], pattern: new RegExp("") },
    modules: [],
    loaded: false,
    elements: {},
  },

  isFull: false,
  location: "",
  params: {},
});

export const MatchProvider = MatchContext.Provider;

export function useMatch(): Match {
  return useContext(MatchContext);
}
