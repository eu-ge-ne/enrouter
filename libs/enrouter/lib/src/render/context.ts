import { createContext } from "react";

import type { RouteNodes } from "./mod.js";

export const RouteRenderContext = createContext<RouteNodes>({
  match: {
    handler: {
      route: {
        path: "",
        mod: [],
        link: [[], []],
      },
      test: { keys: [], pattern: new RegExp("") },
      modules: [],
    },
    location: "",
    isFull: false,
    params: {},
  },
});
