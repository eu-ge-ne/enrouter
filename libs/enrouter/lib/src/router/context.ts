import { createContext, useContext } from "react";

import type { Route } from "#lib/route/mod.js";

export interface TRouterContext {
  routes: Route;
  location: string;
  navigate: (to: string) => void;
  ctx: unknown;
}

const RouterContext = createContext<TRouterContext>({
  routes: {
    path: "",
    test: { keys: [], pattern: new RegExp("") },
    modules: [],
    loaded: false,
    elements: {},
  },
  location: "",
  navigate: () => undefined,
  ctx: undefined,
});

export const RouterProvider = RouterContext.Provider;

export function useRouter(): TRouterContext {
  return useContext(RouterContext);
}
