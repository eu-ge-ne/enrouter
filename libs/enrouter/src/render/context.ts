import { createContext } from "react";

import type { RouteNodes } from "./mod.js";

export const RouteRenderContext = createContext<RouteNodes>({
  match: {
    params: {},
    handler: {
      route: {
        path: "",
        mod: [],
        link: { css: [], mod: [] },
      },
      test: { keys: [], pattern: new RegExp("") },
      modules: [],
    },
  },
});
