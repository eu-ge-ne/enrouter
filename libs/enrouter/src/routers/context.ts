import { createContext } from "react";

import type { RouteHandler } from "#handlers/mod.js";

export interface TRouterContext {
  handlers: RouteHandler;
  location: string;
  navigate: (to: string) => void;
}

export const RouterContext = createContext<TRouterContext>({
  handlers: {
    route: {
      path: "",
      mod: [],
      link: { css: [], mod: [] },
    },
    test: { keys: [], pattern: new RegExp("") },
    modules: [],
  },
  location: "",
  navigate: () => undefined,
});
