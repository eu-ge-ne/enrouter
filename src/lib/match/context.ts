import { createContext, useContext } from "react";

import type { Match } from "./match.js";

const MatchesContext = createContext<Match[]>([]);
const MatchIndexContext = createContext<number>(-1);

export const MatchesProvider = MatchesContext.Provider;
export const MatchIndexProvider = MatchIndexContext.Provider;

export interface MatchContext {
  matches: Match[];
  matchIndex: number;
  match: Match | undefined;
  firstMatch: Match | undefined;
  nextMatch: Match | undefined;
  lastMatch: Match | undefined;
  voidMatch: Match | undefined;
  isExactMatch: boolean;
}

export function useMatchContext(): MatchContext {
  const matches = useContext(MatchesContext);
  const matchIndex = useContext(MatchIndexContext);

  const match = matches[matchIndex];
  const firstMatch = matches[0];
  const nextMatch = matches[matchIndex + 1];
  const lastMatch = matches.at(-1);
  const voidMatch = matches.findLast((x) => x.route?.elements._void);

  const isExactMatch = Boolean(lastMatch?.route);

  return {
    matches,
    matchIndex,
    match,
    firstMatch,
    nextMatch,
    lastMatch,
    voidMatch,
    isExactMatch,
  };
}

export function useMatch(path: string): Match | undefined {
  return useContext(MatchesContext).find((x) => x.route?.path === path);
}
