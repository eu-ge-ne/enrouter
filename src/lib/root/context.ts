import { type ComponentType, createContext, useContext } from "react";

export interface RootParams {
  root?: ComponentType;
  fallback?: Record<string, ComponentType>;
}

const RootParamsContext = createContext<RootParams>({});

export const RootParamsProvider = RootParamsContext.Provider;

export function useRootParams(): RootParams {
  return useContext(RootParamsContext);
}
