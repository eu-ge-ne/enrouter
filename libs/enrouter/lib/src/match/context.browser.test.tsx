import type { FC, PropsWithChildren } from "react";
import { describe, test, expect } from "vitest";
import { render } from "vitest-browser-react";
import * as regexparam from "regexparam";

import type { Match } from "./mod.js";
import { MatchProvider, useMatch } from "./context.js";

const wrapperId = "test-wrapper";

const wrapper: FC<PropsWithChildren> = ({ children }) => (
  <div data-testid={wrapperId}>{children}</div>
);

describe("match", () => {
  describe("useMatch", () => {
    test("1 match", async () => {
      const context: Match = {
        route: {
          path: "/",
          test: regexparam.parse("/", true),
          modules: [],
          loaded: true,
          elements: {
            layout: {
              Main: <div>Layout /</div>,
            },
          },
        },
        isFull: true,
        location: "/",
        params: {},
      };

      const Test: FC = () => {
        const match = useMatch();
        return match?.route.elements.layout?.Main;
      };

      const screen = render(
        <MatchProvider value={context}>
          <Test />
        </MatchProvider>,
        {
          wrapper,
        },
      );

      await expect.element(screen.getByTestId(wrapperId)).toBeVisible();

      expect(screen.container).toMatchSnapshot();
    });
  });
});
