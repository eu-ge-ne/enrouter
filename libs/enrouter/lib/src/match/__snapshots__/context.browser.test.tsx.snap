// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`match > context > useMatch 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      Layout /
    </div>
  </div>
</div>
`;
